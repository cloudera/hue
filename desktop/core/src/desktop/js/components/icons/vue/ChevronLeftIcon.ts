import { defineComponent } from 'vue';

export default defineComponent({
  name: 'ChevronLeftIcon',
  template: '<svg class="hi hi-fw"><use href="#hueChevronLeftSymbol" /></svg>'
});
