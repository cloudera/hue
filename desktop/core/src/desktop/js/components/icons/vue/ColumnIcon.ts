import { defineComponent } from 'vue';

export default defineComponent({
  name: 'ColumnIcon',
  template: '<svg class="hi hi-fw"><use href="#hueColumnSymbol" /></svg>'
});
