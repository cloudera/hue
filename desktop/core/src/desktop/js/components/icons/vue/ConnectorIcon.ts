import { defineComponent } from 'vue';

export default defineComponent({
  name: 'ConnectorIcon',
  template: '<svg class="hi hi-fw"><use xlink:href="#hueConnectorSymbol" /></svg>'
});
