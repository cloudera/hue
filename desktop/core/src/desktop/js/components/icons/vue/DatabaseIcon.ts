import { defineComponent } from 'vue';

export default defineComponent({
  name: 'DatabaseIcon',
  template: '<svg class="hi hi-fw"><use xlink:href="#hueDatabaseSymbol" /></svg>'
});
