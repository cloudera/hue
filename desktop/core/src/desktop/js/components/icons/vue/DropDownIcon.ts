import { defineComponent } from 'vue';

export default defineComponent({
  name: 'DropDownIcon',
  template: '<svg class="hi hi-fw"><use href="#hueDropDownSymbol" /></svg>'
});
