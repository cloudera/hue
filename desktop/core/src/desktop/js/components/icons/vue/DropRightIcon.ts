import { defineComponent } from 'vue';

export default defineComponent({
  name: 'DropRightIcon',
  template: '<svg class="hi hi-fw"><use href="#hueDropRightSymbol" /></svg>'
});
