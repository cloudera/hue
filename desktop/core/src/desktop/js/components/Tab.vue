<!--
  Licensed to Cloudera, Inc. under one
  or more contributor license agreements.  See the NOTICE file
  distributed with this work for additional information
  regarding copyright ownership.  Cloudera, Inc. licenses this file
  to you under the Apache License, Version 2.0 (the
  "License"); you may not use this file except in compliance
  with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
-->

<template>
  <div v-if="!lazy || rendered" v-show="isActive">
    <slot />
  </div>
</template>

<script lang="ts">
  import { defineComponent, inject } from 'vue';
  import { SELECTED_TAB_KEY, ADD_TAB_KEY, REMOVE_TAB_KEY } from './Tabs.vue';

  export interface TabRef {
    title: string;
  }

  export default defineComponent({
    name: 'Tab',

    props: {
      title: {
        type: String,
        required: true
      },
      lazy: {
        type: Boolean,
        default: false
      }
    },

    setup(props) {
      const addTab: ((tab: TabRef) => void) | undefined = inject(ADD_TAB_KEY);
      const removeTab: ((tab: TabRef) => void) | undefined = inject(REMOVE_TAB_KEY);

      const selectedTabRef: TabRef | undefined = inject(SELECTED_TAB_KEY);

      return {
        addTab,
        removeTab,
        selectedTabRef,

        ref: {
          title: props.title
        }
      };
    },

    data() {
      return {
        rendered: false
      };
    },

    computed: {
      isActive(): boolean {
        return this.ref.title === this.selectedTabRef?.title;
      }
    },

    watch: {
      selectedTabRef(): void {
        if (this.isActive) {
          this.rendered = true;
        }
      }
    },

    mounted(): void {
      if (this.addTab) {
        this.addTab(this.ref);
      }
    }

    // TODO
    // destroyed(): was deprecated, need to rearchitect the component.
    // Whenever parent is rendered mount, unmount is called causing an to prevent infinit loop.
    // unmounted(): void {
    //   if (this.removeTab) {
    //     this.removeTab(this.def);
    //   }
    // }
  });
</script>

<style lang="scss" scoped></style>
