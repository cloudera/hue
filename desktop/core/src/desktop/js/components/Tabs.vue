<!--
  Licensed to Cloudera, Inc. under one
  or more contributor license agreements.  See the NOTICE file
  distributed with this work for additional information
  regarding copyright ownership.  Cloudera, Inc. licenses this file
  to you under the Apache License, Version 2.0 (the
  "License"); you may not use this file except in compliance
  with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
-->

<template>
  <div class="hue-tabs">
    <ul>
      <li
        v-for="tab in tabs"
        :key="tab.title"
        :class="{ active: selectedTabRef === tab }"
        @click="selectedTabRef = tab"
      >
        {{ tab.title }}
      </li>
    </ul>
    <div class="hue-tab-container">
      <slot />
    </div>
  </div>
</template>

<script lang="ts">
  import { defineComponent, ref, provide } from 'vue';

  import { TabRef } from './Tab.vue';

  export const ADD_TAB_KEY = 'addTab';
  export const REMOVE_TAB_KEY = 'removeTab';

  export const SELECTED_TAB_KEY = 'selectedTabRef';

  export default defineComponent({
    name: 'Tabs',

    setup() {
      const tabs = ref<TabRef[]>([]);
      const selectedTabRef = ref<TabRef | null>(null);

      function addTab(tab: TabRef): void {
        if (!tabs.value.find(t => t.title === tab.title)) {
          tabs.value.push(tab);
          if (tabs.value.length === 1) {
            selectedTabRef.value = tabs.value[0];
          }
        }
      }

      function removeTab(tab: TabRef): void {
        const index = tabs.value.indexOf(tab);
        if (index !== -1) {
          tabs.value.splice(index, 1);
          if (tab === selectedTabRef.value && tabs.value.length) {
            selectedTabRef.value = tabs.value[Math.max(0, index - 1)];
          }
        }
      }

      provide(ADD_TAB_KEY, (tab: TabRef) => addTab(tab));
      provide(REMOVE_TAB_KEY, (tab: TabRef) => removeTab(tab));

      provide(SELECTED_TAB_KEY, selectedTabRef);

      return {
        tabs,
        selectedTabRef
      };
    }
  });
</script>

<style lang="scss" scoped>
  @import './styles/colors';

  .hue-tabs {
    ul {
      border-bottom: 1px solid $hue-border-color;
      list-style: none;
      margin: 0;

      li {
        display: inline-block;
        font-size: 14px;
        padding: 10px;
        border-bottom: 3px solid transparent;
        cursor: pointer;
        margin-right: 10px;
        color: $fluid-gray-700;

        &:hover {
          color: $fluid-blue-500;
        }

        &.active {
          color: $fluid-blue-500;
          border-bottom: 3px solid $fluid-blue-400;
        }
      }
    }
    .hue-tab-container {
      position: relative;
      margin: 12px 0;
      background-color: $fluid-white;
    }
  }
</style>
