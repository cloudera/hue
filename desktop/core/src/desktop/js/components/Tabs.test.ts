// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// 'License'); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an 'AS IS' BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { nextTick } from 'vue';
import { mount, shallowMount } from '@vue/test-utils';
import Tabs from './Tabs.vue';
import Tab from './Tab.vue';

describe('Tabs.vue', () => {
  it('should render empty tabs', () => {
    const wrapper = shallowMount(Tabs);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render tabs', async () => {
    const wrapper = mount(Tabs, {
      slots: {
        default: '<tab title="foo">foo</tab><tab title="bar">bar</tab>'
      },
      global: {
        stubs: {
          Tab
        }
      }
    });

    await nextTick();
    expect(wrapper.element).toMatchSnapshot();
  });
});
