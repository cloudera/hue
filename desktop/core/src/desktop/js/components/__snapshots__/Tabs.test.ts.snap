// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs.vue should render empty tabs 1`] = `
<div
  class="hue-tabs"
>
  <ul>
    
    
  </ul>
  <div
    class="hue-tab-container"
  >
    
    
  </div>
</div>
`;

exports[`Tabs.vue should render tabs 1`] = `
<div
  class="hue-tabs"
>
  <ul>
    
    <li
      class="active"
    >
      foo
    </li>
    <li
      class=""
    >
      bar
    </li>
    
  </ul>
  <div
    class="hue-tab-container"
  >
    
    
    <div>
      
      foo
      
    </div>
    <div
      style="display: none;"
    >
      
      bar
      
    </div>
    
    
  </div>
</div>
`;
